geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/process/RouteSearchProcess.scala

```diff
@@ -0,0 +1,354 @@
+    *
+    * @param source simple feature source
+    * @param query may contain additional filters to apply
+    */
+  def routeSearch(source: SimpleFeatureSource, query: Query): Unit = {
+    logger.debug(s"Visiting source type: ${source.getClass.getName}")
+    val result = if (routes.isEmpty) { source.getFeatures(Filter.EXCLUDE) } else {
+      import org.locationtech.geomesa.filter._
+      val filter = if (query == null || query.getFilter == Filter.INCLUDE) { routeFilter } else {
+        andFilters(Seq(query.getFilter, routeFilter))
+      }
+      matchRoutes(source.getFeatures(filter))
+    }
+    resultCalc = RouteResult(result)
+  }
+
+  /**
+    * Filters a feature collection by comparing routes
+    *
+    * @param input input feature collection
+    * @return
+    */
+  private def matchRoutes(input: SimpleFeatureCollection): SimpleFeatureCollection = {
+    val sft = input.getSchema
+    val geomIndex = sft.indexOf(geomAttribute)
+    val headingIndex = headingAttribute.map(sft.indexOf)
+
+    logger.debug(s"Searching routes: ${routes.map(WKTUtils.write).mkString(", ")}")
+    logger.debug(s"Buffer (meters): $routeBuffer")
+    logger.debug(s"Geometry attribute: $geomAttribute")
+    logger.debug(s"Heading attribute: ${headingAttribute.getOrElse("none")}")
+
+    RouteVisitor.matchRoutes(input, routes, geomIndex, isPoints, headingIndex, threshold, bidirectional)
+  }
+}
+
+object RouteVisitor {
+
+  /**
+    * Match features to routes.
+    *
+    * @param input input features
+    * @param routes routes to match against
+    * @param geomIndex geometry attribute index of the input features
+    * @param isPoints are the input feature geometries points (or linestrings)
+    * @param headingIndex attribute index of the heading in the input features, not required if they are linestrings
+    * @param threshold threshold to consider when matching route heading, in degrees
+    * @param bidirectional match the route direction or just the path
+    * @return
+    */
+  def matchRoutes(input: SimpleFeatureCollection,
+                  routes: Seq[LineString],
+                  geomIndex: Int,
+                  isPoints: Boolean,
+                  headingIndex: Option[Int],
+                  threshold: Double,
+                  bidirectional: Boolean): SimpleFeatureCollection = {
+
+    // just in case the features are operated on in parallel...
+    val calculator = new ThreadLocal[GeodeticCalculator] {
+      override def initialValue(): GeodeticCalculator = new GeodeticCalculator
+    }
+    val headingCache = {
+      import scala.collection.JavaConverters._
+      new ConcurrentHashMap[(LineString, Int), Double].asScala
+    }
+
+    // gets the heading for an input feature
+    val getFeatureHeading: (SimpleFeature) => Double = headingIndex match {
+      case Some(index) => (sf) => Converters.convert(sf.getAttribute(index), classOf[Double])
```

Comment:
it does type casting, so if the type was actually a string, or an int, or whatever it would work where the cast would fail. It may be overkill but it's handy